src/client/Inode.cc

```diff
@@ -552,3 +560,146 @@ void Inode::set_async_err(int r)
   }
 }
 
+bool Inode::has_recalled_deleg()
+{
+  if (delegations.empty())
+    return false;
+
+  Delegation *deleg = delegations.front();
+
+  return deleg->is_recalled();
+}
+
+void Inode::recall_deleg(bool skip_read)
+{
+  if (delegations.empty())
+    return;
+
+  // Issue any recalls
+  for (xlist<Delegation*>::iterator p = delegations.begin(); !p.end(); ++p) {
+    Delegation *deleg = *p;
+
+    deleg->recall(skip_read);
+  }
+}
+
+bool Inode::delegations_broken(bool skip_read)
+{
+  if (delegations.empty())
+    return true;
+
+  if (skip_read) {
+    Delegation *deleg = delegations.front();
+    if (deleg->get_mode() == CEPH_FILE_MODE_RD)
+	return true;
+  }
+  return false;
+}
+
+void Inode::break_deleg(bool skip_read)
+{
+  recall_deleg(skip_read);
+
+  while (!delegations_broken(skip_read))
+    client->wait_on_list(waitfor_deleg);
+}
+
+/**
+ * set_deleg: request a delegation on an open Fh
+ * @fh: filehandle on which to acquire it
+ * @ro: r/o or r/w delegation?
+ * @cb: delegation recall callback function
+ * @priv: private pointer to be passed to callback
+ *
+ * Attempt to acquire a delegation on an open file handle. If there are no
+ * conflicts and we have the right caps, allocate a new delegation, fill it
+ * out and return 0. Return an error if we can't get one for any reason.
+ */
+int Inode::set_deleg(Fh *fh, bool ro, ceph_deleg_cb_t cb, void *priv)
+{
+  int filemode;
+
+  lsubdout(client->cct, client, 0) << __func__ << ": inode " << *this << dendl;
+
+  // Just say no if we have any recalled delegs still outstanding
+  if (has_recalled_deleg()) {
+    lsubdout(client->cct, client, 0) << __func__ << ": has_recalled_deleg" << dendl;
+    return -EAGAIN;
+  }
+
+  // check vs. currently open files on this inode
+  if (ro) {
+    if (open_count_for_write()) {
+      lsubdout(client->cct, client, 0) << __func__ << ": open for write" << dendl;
+      return -EAGAIN;
+    }
+    filemode = CEPH_FILE_MODE_RD;
+  } else {
+    if (open_count() > 1) {
+      lsubdout(client->cct, client, 0) << __func__ << ": open" << dendl;
+      return -EAGAIN;
+    }
+    filemode = CEPH_FILE_MODE_RDWR;
+  }
+
+  // Do we have the caps we need for this inode?
+  int need = ceph_caps_for_mode(filemode);
+  if (!caps_issued_mask(need)) {
+    lsubdout(client->cct, client, 0) << __func__ << ": cap mismatch, have="
+      << ccap_string(caps_issued()) << " need=" << ccap_string(need) << dendl;
+    return -EAGAIN;
+  }
+
+  Delegation *deleg, *old = NULL;
+  for (xlist<Delegation*>::iterator p = delegations.begin(); !p.end(); ++p) {
+    deleg = *p;
+    if (deleg->get_fh() == fh) {
+      old = deleg;
+      break;
+    }
+  }
+
+  if (old) {
+    unsigned oldmode = old->get_mode();
+
+    client->get_cap_ref(this, need);
+    old->reinit(filemode, cb, priv);
+    client->put_cap_ref(this, ceph_caps_for_mode(oldmode));
+  } else {
+    deleg = new (std::nothrow) Delegation(fh, filemode, cb, priv);
```

Comment:
Sigh...I don't understand the hate for xlist here. If you use std::list then you:

a/ have to do a second allocation every time (to allocate the list container). That's nasty because code is part of a library. It's bad form to crash the program on an allocation failure (no matter how unlikely), particularly in something where we have a very simple recovery path (just not granting a delegation in this case). I suppose I could try to catch that with an exception, but this is far simpler.

b/ it also means that when you want to remove an object from the list, that you must walk the whole damn list until you find the object. An embedded list head (like xlist) gives you the ability to remove it in constant time with just some pointer math.

Finally, xlists are used all over the client code. If you really want to standardize on another sort of list, then so be it, but I don't see a reason to avoid it here.

Reply:
> Sigh...I don't understand the hate for xlist here.

It's not hate for xlist, just for the further proliferation of pointer managed objects. Besides simplifying memory management and avoiding potential leaks in the future, it also makes a future conversion to a mempool managed Client cache easier (similar to what was done for the MDS).

> a/ have to do a second allocation every time (to allocate the list container).

I assume you mean if we do `std::list<Delegation *>` where we need to allocated the `Delegation` and also push onto the list (a second allocation)? I'm recommending `std::list<Delegation>`. It's all one allocation for each new Delegation.

> b/ it also means that when you want to remove an object from the list, that you must walk the whole damn list until you find the object.

We're already iterating the whole list on removal in `unset_deleg`. For that reason, it now seems to me that a `std::map<Fh *, Delegation>` would be more applicable.